crates/runner/src/lib.rs

```diff
@@ -74,88 +283,208 @@ pub fn run_cairo_program(
+fn encode_value_for_call(
+    vm: &mut VM,
+    array_cursor: &mut M31,
+    ty: &AbiType,
+    val: &InputValue,
+    dst: &mut Vec<M31>,
+) -> Result<()> {
+    match (ty, val) {
+        (AbiType::Felt, InputValue::Number(n)) => dst.push(m31_from_i64(*n)),
+        (AbiType::Bool, InputValue::Number(n)) => match *n {
+            0 => dst.push(M31::from(0u32)),
+            1 => dst.push(M31::from(1u32)),
+            _ => {
+                return Err(
+                    AbiCodecError::TypeMismatch(format!("bool expects 0 or 1, got {}", n)).into(),
+                )
+            }
+        },
+        (AbiType::Bool, InputValue::Bool(b)) => dst.push(M31::from(if *b { 1u32 } else { 0u32 })),
+        (AbiType::U32, InputValue::Number(n)) => {
+            if *n < 0 || (*n as i128) > u32::MAX as i128 {
+                return Err(AbiCodecError::TypeMismatch(format!("u32 out of range: {}", n)).into());
+            }
+            let u = *n as u32;
+            let lo = M31::from(u & 0xFFFF);
+            let hi = M31::from(u >> 16);
+            dst.extend_from_slice(&[lo, hi]);
+        }
+        (AbiType::Pointer(_), _) => {
+            // Pointers cannot be provided directly as inputs in runner
+            return Err(AbiCodecError::TypeMismatch(
+                "Pointer types are internal-only and cannot be provided as input.".to_string(),
+            )
+            .into());
+        }
+        (AbiType::Tuple(types), InputValue::List(values)) => {
+            if types.len() != values.len() {
+                return Err(AbiCodecError::TypeMismatch(format!(
+                    "tuple arity mismatch: expected {} got {}",
+                    types.len(),
+                    values.len()
+                ))
+                .into());
+            }
+            for (t, v) in types.iter().zip(values.iter()) {
+                encode_value_for_call(vm, array_cursor, t, v, dst)?;
+            }
+        }
+        (AbiType::Struct { fields, .. }, InputValue::Struct(values)) => {
+            if fields.len() != values.len() {
+                return Err(AbiCodecError::TypeMismatch(format!(
+                    "struct field count mismatch: expected {} got {}",
+                    fields.len(),
+                    values.len()
+                ))
+                .into());
+            }
+            for ((_, fty), v) in fields.iter().zip(values.iter()) {
+                encode_value_for_call(vm, array_cursor, fty, v, dst)?;
+            }
+        }
+        (AbiType::FixedSizeArray { element, size }, InputValue::List(values)) => {
+            if *size as usize != values.len() {
+                return Err(AbiCodecError::TypeMismatch(format!(
+                    "array size mismatch: expected {} got {}",
+                    size,
+                    values.len()
+                ))
+                .into());
+            }
+
+            // First, encode each element into a temporary buffer using the *argument ABI*.
```

Comment:
this method will be used in the hot pass (run_cairo_program) and temporary buffers / unecessary allocations should be avoided.

Reply:
Sure but there are 2 pass and create a temporary vec `elements_m31` that could be avoided by doing the encoding and writing to memory in only one pass.

Reply:
I don't see how you can avoid two loops here